### PR TITLE
Daily job to generate dashboards for released BOM versions

### LINKED ISSUE #158: Multiple Version dashboard
The dashboard should work across multiple versions so we can measure and show progress as the cloud-oss-bom evolves.

Maybe figure out how to track metrics of some kind?

### LINKED ISSUE #697: dashboard.html --> index.html
in dashboard

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #158 .
Fixes #697.

- Using Maven Central's API to list versions
- Re-using the existing Kokoro job configuration to upload GCS
- Renaming dashboard.html to groupId/artifactId/version/index.html

Example screenshot. version:1.0.0, 1.1.1 and the latest snapshot
image

### CODE REVIEW

File: kokoro/ubuntu/periodic.sh [hunk 1]
```diff
@@ -13,5 +13,22 @@ mvn -B clean install
 # https://stackoverflow.com/questions/3459928/running-a-specific-maven-plugin-goal-from-the-command-line-in-a-sub-module-of-a/26448447#26448447
 # https://stackoverflow.com/questions/11091311/maven-execjava-goal-on-a-multi-module-project
 cd dashboard
+
+# Step 1: Generate dashboards for the released BOMs with the latest DashboardMain class
+# For example https://storage.googleapis.com/cloud-opensource-java-dashboard/dashboard/target/1.1.0/dashboard.html
+
+# Querying https://search.maven.org/classic/#api with
+# groupId:com.google.cloud AND artifactId:libraries-bom, receiving latest (sorted) 10 versions.
```
Comment (suztomo) [inline kokoro/ubuntu/periodic.sh]: The document (https://search.maven.org/classic/#api) says it's sorted.

> Mimics clicking the link for all versions of groupId "com.google. inject" and artifactId "guice." Returns sorted list of all versions of an artifact.

File: kokoro/ubuntu/periodic.sh [hunk 2]
```diff
@@ -13,5 +13,22 @@ mvn -B clean install
 # https://stackoverflow.com/questions/3459928/running-a-specific-maven-plugin-goal-from-the-command-line-in-a-sub-module-of-a/26448447#26448447
 # https://stackoverflow.com/questions/11091311/maven-execjava-goal-on-a-multi-module-project
 cd dashboard
+
+# Step 1: Generate dashboards for the released BOMs with the latest DashboardMain class
+# For example https://storage.googleapis.com/cloud-opensource-java-dashboard/dashboard/target/1.1.0/dashboard.html
```
Comment (suztomo) [inline kokoro/ubuntu/periodic.sh]: I feel "dashboard/target" in the URL is unnecessary. I think it can be removed by updating Kokoro job config

```
strip_prefix: "github/cloud-opensource-java/"
```
to
```
strip_prefix: "github/cloud-opensource-java/dashboard/target"
```

@elharo What do you think?

File: kokoro/ubuntu/periodic.sh [hunk 3]
```diff
@@ -13,5 +13,22 @@ mvn -B clean install
...
+# Example: '1.0.0 1.1.0'
+VERSIONS=`echo $SONATYPE_RESPONSE | perl -nle 'print $1 while m/"v":"(.+?)"/g'`
+for VERSION in $VERSIONS; do
+  mvn -B exec:java -Dexec.mainClass="com.google.cloud.tools.opensource.dashboard.DashboardMain" \
+    -Dexec.arguments="-c com.google.cloud:libraries-bom:${VERSION}"
+  mv target/dashboard target/${VERSION}
+done
+
+# Step 2: Generate dashboard for the snapshot
+# For example https://storage.googleapis.com/cloud-opensource-java-dashboard/dashboard/target/dashboard/dashboard.html
```
Comment (suztomo) [inline kokoro/ubuntu/periodic.sh]: `dashboard/target/dashboard` seems giving no value. With Kokoro job config (see above) and one `mv` command here, the URL for the snapshot BOM can become 

```
https://storage.googleapis.com/cloud-opensource-java-dashboard/snapshot/dashboard.html
```

File: kokoro/ubuntu/periodic.sh [hunk 4]
```diff
@@ -13,5 +13,23 @@ mvn -B clean install
...
 # https://stackoverflow.com/questions/11091311/maven-execjava-goal-on-a-multi-module-project
 cd dashboard
+
+# Step 1: Generate dashboards for the released BOMs with the latest DashboardMain class
+# For example: https://storage.googleapis.com/cloud-opensource-java-dashboard/1.1.0/dashboard.html
+
+# Querying https://search.maven.org/classic/#api with
+# groupId:com.google.cloud AND artifactId:libraries-bom, receiving latest (sorted) versions.
+# Note: if `rows` parameter is omitted, it only returns latest 10 items.
+SONATYPE_RESPONSE=`curl 'https://search.maven.org/solrsearch/select?q=g:%22com.google.cloud%22+AND+a:%22libraries-bom%22&core=gav&rows=1000&wt=json'`
```
Comment (suztomo) [inline kokoro/ubuntu/periodic.sh]: Default value for rows parameter is only 10 (example below). So explicitly specifying the value here.
Example with guava
https://search.maven.org/solrsearch/select?q=g:%22com.google.guava%22+AND+a:%22guava%22&core=gav&wt=json

Comment (elharo) [reply kokoro/ubuntu/periodic.sh]: Ideally we should page this.

Comment (suztomo) [reply kokoro/ubuntu/periodic.sh]: YAGNI

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 5]
```diff
@@ -393,4 +393,10 @@ public void testGlobalUpperBoundUpgradeMessage() throws IOException, ParsingExce
     Truth.assertThat(dependencyUpgradeMessage).contains(
         "Upgrade com.google.guava:guava:jar:19.0 to version \"27.1-android\"");
   }
+
+  @Test
+  public void testOutputDirectory() {
+    Truth.assertThat(outputDirectory.toString())
+        .isEqualTo("target/com.google.cloud/libraries-bom/1.0.0");
```
Comment (suztomo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: Does it work with Windows?

Comment (suztomo) [reply dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: It didn't. Updated.

**Comment by suztomo** [comment]
@elharo PTAL.

File: DECLARING_DEPENDENCIES.md [hunk 6]
```diff
@@ -181,7 +181,8 @@ document. Such conflicts are called *intrinsic conflicts*. There is an
 ongoing effort to remove intrinsic conflicts among GCP open source
 Java libraries and prevent new ones from occurring. A dashboard that
 reports the current results of compatibility checks is accessible from
```
Comment (elharo) [inline DECLARING_DEPENDENCIES.md]: "A dashboard that reports the current results of compatibility checks is accessible from the [Cloud Open Source Java Dashboard]" --> "The [Cloud Open Source Java Dashboard] reports the current results of compatibility checks."

File: README.md [hunk 7]
```diff
@@ -6,7 +6,7 @@ Java projects for the Google Cloud Platform.
 # Google Cloud Platform Java Dependency Dashboard
 
 [Google Cloud Platform Java Dependency Dashboard](
-https://storage.googleapis.com/cloud-opensource-java-dashboard/dashboard/dashboard.html)
+https://storage.googleapis.com/cloud-opensource-java-dashboard/com.google.cloud/libraries-bom/snapshot/index.html)
```
Comment (elharo) [inline README.md]: Probably doesn't matter a great deal, but this maybe should be a separate PR just for docs since this URL won't be immediately live when you commit and the docs will be.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardArguments.java [hunk 8]
```diff
@@ -62,7 +62,7 @@ Path getBomFile() {
   /** Returns the Maven coordinates of a BOM. Null if coordinates are not specified. */
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardArguments.java]: looks like now a null pointer exception if coordinates aren't specified

File: kokoro/ubuntu/periodic.sh [hunk 9]
```diff
@@ -13,5 +13,25 @@ mvn -B clean install
...
+
+# Step 1: Generate dashboards for the released BOMs with the latest DashboardMain class
+# For example: https://storage.googleapis.com/cloud-opensource-java-dashboard/com.google.cloud/libraries-bom/1.1.1/index.html
+
+# Querying https://search.maven.org/classic/#api with
+# groupId:com.google.cloud AND artifactId:libraries-bom, receiving latest (sorted) 1000 versions.
+# Note: if `rows` parameter is omitted, it only returns latest 10 items.
+SONATYPE_RESPONSE=`curl 'https://search.maven.org/solrsearch/select?q=g:%22com.google.cloud%22+AND+a:%22libraries-bom%22&core=gav&rows=1000&wt=json'`
+# Example: '1.0.0 1.1.0'
+VERSIONS=`echo $SONATYPE_RESPONSE | perl -nle 'print $1 while m/"v":"(.+?)"/g'`
```
Comment (elharo) [inline kokoro/ubuntu/periodic.sh]: How reliable is parsing JSON with regexps? What if Sonatype reformats? How are errors reported?
